### ISSUE TITLE
`SuccessMessageMixin` does not allow `form_valid` to be retyped

### ISSUE THREAD

**Issue by Kakadus** [typeddjango/django-stubs#1809]
# Bug report

## What's wrong

When using the `SuccessMessageMixin`, it is impossible to retype `form_valid`, because it is already defined by `SuccessMessage`. #86 seems somewhat related, however, this bug is about the `form` argument.

```python
from django import forms
from django.contrib.messages.views import SuccessMessageMixin
from django.forms import ModelForm


class MyForm(forms.Form):
    pass


class Test(SuccessMessageMixin, ModelForm):
    def form_valid(self, form: MyForm):
        pass
```
This results in 
> evap/staff/views.py:1068: error: Argument 1 of "form_valid" is incompatible with supertype "SuccessMessageMixin"; supertype defines the argument type as "BaseForm"  [override]  
evap/staff/views.py:1068: note: This violates the Liskov substitution principle  
evap/staff/views.py:1068: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides  
Found 1 error in 1 file (checked 122 source files)

## How it should be

It should be possible to narrow the type of `field` in `form_valid`. It should be possible, if `SuccessMessageMixin` becomes generic to supply the type of the filed:

```python
from django import forms
from django.contrib.messages.views import SuccessMessageMixin
from django.forms import ModelForm


class MyForm(forms.Form):
    pass


class Test(SuccessMessageMixin[MyForm], ModelForm):
    def form_valid(self, form: MyForm):
        pass
```

## System information

- OS: Arch Linux
- `python` version: 3.10
- `django` version: 4.2.0
- `mypy` version: 1.6.1
- `django-stubs` version: 4.2.5
- `django-stubs-ext` version: 4.2.5

**Comment by intgr** [comment]
Makes sense. Are you interested in contributing a fix?

**Comment by Kakadus** [comment]
Sure! I'm a bit short on time, but will be able to open a PR on Monday next week

**Comment by SukiCZ** [comment]
Sorry @Kakadus to jump in. I saw a "good first issue" label and decided to fix it in PR #1812 
Thanks for a nice bug report 😻  It helped me a lot!
